File: examples/spookyball/src/jsMain/kotlin/transform.kt
import org.khronos.webgl.ArrayBuffer
import org.khronos.webgl.Float32Array

val DEFAULT_ORIENTATION = Float32Array(arrayOf(0f, 0f, 0f, 0f))

@OptIn(ExperimentalJsExport::class)
@JsExport
val DEFAULT_POSITION = MyVector3(0.0, 0.0, 0.0)

@OptIn(ExperimentalJsExport::class)
@JsExport
val DEFAULT_SCALE = MyVector3(1.0, 1.0, 1.0)

@OptIn(ExperimentalJsExport::class)
@JsExport
class TransformKt(options: dynamic) {

    var position: Float32Array
    var orientation: Float32Array
    var scale: Float32Array
    var localMatrix: Float32Array
    var worldMatrix: Float32Array
    var localMatrixDirty = true
    var worldMatrixDirty = true

    init {

        var buffer: ArrayBuffer
        var offset = 0

        if (options.externalStorage != null) {
            buffer = options.externalStorage.buffer
            offset = options.externalStorage.offset
        } else {
            buffer = Float32Array(42).buffer
        }

        position = Float32Array(buffer, offset, 3)
        orientation = Float32Array(buffer, offset + 3 * Float32Array.BYTES_PER_ELEMENT, 4)
        scale = Float32Array(buffer, offset + 7 * Float32Array.BYTES_PER_ELEMENT, 3)
        localMatrix = Float32Array(buffer, offset + 10 * Float32Array.BYTES_PER_ELEMENT, 16)
        worldMatrix = Float32Array(buffer, offset + 26 * Float32Array.BYTES_PER_ELEMENT, 16)

        if (options.transform) {
            val storage = Float32Array(position.buffer, position.byteOffset, 42)
            storage.set(Float32Array(options.transform.actual.position.buffer, options.transform.actual.position.byteOffset, 42))
            localMatrixDirty = options.transform.localMatrixDirty;
        } else {
            if (options.position) {
                position.set(options.position as Array<Float>)
            }
            orientation.set(if(options.orientation) options.orientation as Float32Array else DEFAULT_ORIENTATION)
            scale.set(if(options.scale) options.scale as Float32Array else DEFAULT_SCALE.toJS32Array())
        }

        if (options.parent) {
            options.parent.addChild(this);
        }
    }

}

File: examples/spookyball/src/jsMain/resources/js/engine/core/transform.js
import {
    mat4FromRotationTranslationScale,
    mat4Multiply,
    MyVector3,
    TransformKt,
    vec3TransformMat4
} from "../../spookyball.js";


export class Transform {
    actual

    #parent = null;
    #children;

    constructor(options = {}) {

        this.actual = new TransformKt(options)


    }

    get position() {
        this.#makeDirty();
        return this.actual.position;
    }

    set position(value) {
        this.#makeDirty();
        this.actual.position.set(value);
    }

    getWorldPosition(out, position) {
        if (position) {
            if (position != out) {
                out.set(position);
            }
        } else {
            new MyVector3(0, 0, 0).into(out);
        }
        vec3TransformMat4(out, out, this.worldMatrix);
    }

    get orientation() {
        this.#makeDirty();
        return this.actual.orientation;
    }

    set orientation(value) {
        this.#makeDirty();
        this.actual.orientation.set(value);
    }

    get scale() {
        this.#makeDirty();
        return this.actual.scale;
    }

    set scale(value) {
        this.#makeDirty();
        this.actual.scale.set(value);
    }

    get worldMatrix() {
        return this.#resolveWorldMatrix();
    }

    addChild(transform) {
        if (transform.parent && transform.parent != this) {
            transform.parent.removeChild(transform);
        }

        if (!this.#children) {
            this.#children = new Set();
        }
        this.#children.add(transform);
        transform.#parent = this;
        transform.#makeDirty(false);
    }

    removeChild(transform) {
        const removed = this.#children?.delete(transform);
        if (removed) {
            transform.#parent = null;
            transform.#makeDirty(false);
        }
    }

    get children() {
        return this.#children?.values() || [];
    }

    get parent() {
        return this.#parent;
    }

    #makeDirty(markLocalDirty = true) {
        if (markLocalDirty) {
            this.actual.localMatrixDirty = true;
        }
        if (this.actual.worldMatrixDirty) {
            return;
        }
        this.actual.worldMatrixDirty = true;

        if (this.#children) {
            for (const child of this.#children) {
                child.#makeDirty(false);
            }
        }
    }

    #resolveLocalMatrix() {
        const wasDirty = this.actual.localMatrixDirty;
        if (wasDirty) {
            mat4FromRotationTranslationScale(this.actual.localMatrix,
                this.actual.orientation,
                this.actual.position,
                this.actual.scale);
            this.actual.localMatrixDirty = false;
        }
        return this.actual.localMatrix;
    }

    #resolveWorldMatrix() {
        if (this.actual.worldMatrixDirty) {
            if (!this.parent) {
                this.actual.worldMatrix.set(this.#resolveLocalMatrix());
            } else {
                mat4Multiply(this.actual.worldMatrix, this.parent.worldMatrix, this.#resolveLocalMatrix());
            }
            this.actual.worldMatrixDirty = false;
        }

        return this.actual.worldMatrix;
    }
}

export class TransformPool {
    #buffer;
    #transforms = [];

    constructor(size) {
        this.#buffer = new Float32Array(42 * size).buffer;

        for (let i = 0; i < size; ++i) {
            this.#transforms[i] = new Transform({
                externalStorage: {
                    buffer: this.#buffer,
                    offset: (i * 42 * Float32Array.BYTES_PER_ELEMENT),
                }
            });
        }
    }

    get size() {
        return this.#transforms.length;
    }

    getTransform(index) {
        return this.#transforms[index];
    }

    clone() {
        const out = new TransformPool(this.size);
        // Copy the entire buffer from this pool to the new one.
        new Float32Array(out.#buffer).set(new Float32Array(this.#buffer));
        return out;
    }
}
